design/authenticators/application-identity/application-identity.md

```diff
@@ -0,0 +1,464 @@
+      authenticator_name: @authenticator_name, 
+      service_id: @service_id, 
+      role_annotations: @host_annotations, 
+      permitted_annotations: @permitted_constraints,
+    )
+  end
+
+  def verify_configuration
+    # verify that `@host_application_identity_class.constraints.keys` contains
+    # all members of `@required_constraints` and do not have more than one
+    # member of `@non_permitted_combinations`
+    #
+    # Note: this can be extracted to another CommandClass 
+  end
+
+  def verify_application_identity_matches_authentication_request
+    # iterate over `@authentication_request_application_identity.constraints` 
+    # and verify each value matches the corresponding one in 
+    # `@host_application_identity_class.constraints`
+    #
+    # Note: this can be extracted to another CommandClass 
+  end
+end
+```
+
+A call to this class will be made in the relevant Authenticator class. For
+ example, the Azure Authenticator class will look roughly like this:
+ 
+```
+module Authentication
+  module AuthnAzure
+
+    Authenticator = CommandClass.new(
+      dependencies: {
+        <other dependencies>,
+        resource_class:             ::Resource
+        authentication_request_application_identity_class: AuthenticationRequestApplicationIdentity
+        validate_application_identity: ValidateApplicationIdentity.new,
+      },
+      inputs:       [:authenticator_input]
+    ) do
+
+      extend Forwardable
+      def_delegators :@authenticator_input, :service_id, :authenticator_name, :account, :credentials, :username
+
+      def call
+        <perform other authentication steps>
+        validate_application_identity
+      end
+
+      private
+      
+      def validate_application_identity
+        @validate_application_identity.call(
+          authenticator_name: @authenticator_input.authenticator_name
+          service_id: @authenticator_input.service_id
+          host_annotations: role.annotations
+          permitted_constraints: %w(
+                                     subscription-id
+                                     resource-group
+                                     user-assigned-identity
+                                     system-assigned-identity
+                                 )
+          required_constraints: %w(
+                                    subscription-id
+                                    resource-group
+                                 )
+          non_permitted_combinations: %w(
+                                         user-assigned-identity
+                                         system-assigned-identity
+                                       )
+          authentication_request_application_identity: @authentication_request_application_identity_class.new(
+            constraints: @token_identity
+          )
+        )
+      end
+
+      def token_identity_from_claims
+        @token_identity = {
+          subscription_id: xms_mirid.subscriptions,
```

Comment:
where does `xms_mirid` come from if not as input?

Reply:
it does but it's not very interesting. I'll add a comment to clarify